### PR TITLE
Improve performance of client directive validation

### LINKED ISSUE #1554: [Swift] Huge performance regression
Going from 2.17.4 to 2.18.3. Codegen time ballooned from 40s to 240s.

`codegen:generate --includes='./**/*.graphql' --localSchemaFile="schema.json" --passthroughCustomScalars --target=swift Sources/Core/Generated/`

**How to reproduce the issue:**

To help understand the size of this project:
87 .graphql files
schema is 136,639 lines
Sources/Core/Generated/ is 2.2MB

**Versions**

2.18.3

cc @designatednerd

### PR DESCRIPTION
**Pull request by jbaxleyiii**
Attempt to fix #1554 

This commit seeks to improve the performance when determining if a field is missing any @client directives when using local state with Apollo Client. Instead of recursing over the entire document for each field, we walk the tree once while checking the schema for any possible client side annotations. This PR is an initial take at improving this workload but doesn't complete the work.

Still to be done includes supporting @client validation on fragment spreads

### CODE REVIEW

**Comment by jhampton** [comment]
Looking good @jbaxleyiii

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,11 +3,12 @@
 ## Upcoming
 
 - `apollo`
+  - Improve performance of CLI when running projects by delaying or not calling validation [#1559](https://github.com/apollographql/apollo-tooling/pull/1559)
```
Comment (lilyball) [inline CHANGELOG.md]: This change note sounds kind of scary. Can we add the word "unnecessarily" at the end to make it obvious that we're omitting validation when when didn't need it?

File: packages/apollo-language-server/src/errors/validation.ts [hunk 2]
```diff
@@ -106,65 +111,148 @@ export function NoTypenameAlias(context: ValidationContext) {
...
-            [node],
-            null,
-            null,
-            null,
-            null,
-            extensions
-          )
+  // this isn't really execution context, but it does group the fragments and operations
+  // together correctly
+  // XXX we have a simplified version of this in @apollo/gateway that we could proably use
```
Comment (lilyball) [inline packages/apollo-language-server/src/errors/validation.ts]: Typo: "proably"

File: CHANGELOG.md (hunk 1 above)
Comment (JakeDawkins) [reply CHANGELOG.md]: Good thought @lilyball! Updated

**Comment by JakeDawkins** [comment]
Tested with multiple projects in vscode as well, and I don't see any regressions here around validation or startup costs (although no noticeable improvement either).

**Pull request merged**